### PR TITLE
Zero-shot image segmentation with CLIPSeg

### PR DESCRIPTION
**Pull request by segments-tobias**
Hi all! I'm back with another blog post written by @NielsRogge and myself.
The blog post aims to do the following:

- Introduce segmentation and zero-shot segmentation
- Explain how CLIP and CLIPSeg work
- Show how people can use CLIPSeg with `transformers` for both text **and image** prompts
- Show how to combine multiple outputs of CLIPSeg and upload this result to a labeling platform ([Segments.ai](segments.ai) in this case) for further refinement

### CODE REVIEW

**Comment by osanseviero** [comment]
@nateraw and @sayakpaul would you be up for reviewing this blog post? :D

**Comment by sayakpaul** [comment]
I will review it :)

**Review by sayakpaul — CHANGES_REQUESTED** [review]
Thanks for the great work! 

I really enjoyed the step-by-step instructions and the bit on programmatic label refinement. 

Some minor comments to address and we should be good to ship this 🚀

File: clipseg-zero-shot.md [hunk 1]
```diff
@@ -0,0 +1,387 @@
+---
+title: Zero-shot image segmentation with CLIPSeg
+thumbnail: /blog/assets/123_clipseg-zero-shot/thumb.png
```
[the remark this answers is not in the document]
Comment (segments-tobias) [reply clipseg-zero-shot.md:3]: Done! [https://huggingface.co/datasets/huggingface/documentation-images/discussions/18](https://huggingface.co/datasets/huggingface/documentation-images/discussions/18)

Do the image urls stay the same, or should I change them?

File: clipseg-zero-shot.md [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+
+The reason why CLIP works so well is that the model was trained on a huge dataset of images with text captions. The dataset contained a whopping 400 million image-text pairs taken from the internet. These images contain a wide variety of objects and concepts, and CLIP is great at creating a representation for each of them.
+
+## CLIPSeg: image segmentation with CLIP
+
+[CLIPSeg](https://arxiv.org/abs/2112.10003) is a model that uses CLIP representations to create image segmentation masks. It was published by Timo Lüddecke and Alexander Ecker. They achieved zero-shot image segmentation by training a Transformer-based decoder on top of the CLIP model, which is kept frozen. The decoder takes in the CLIP representation of an image, and the CLIP representation of the the thing you want to segment. Using these two inputs, the CLIPSeg decoder creates a binary segmentation mask. To be more precise, the decoder doesn’t only use the final CLIP representation of the image we want to segment, but it also uses the outputs of some of the layers of CLIP.
```
[the remark this answers is not in the document]
Comment (segments-tobias) [reply clipseg-zero-shot.md:66]: So as I try to explain two paragraphs later, you can either input a piece of text (e.g. "an orange"), or a picture (of an orange) to segment oranges in a picture (e.g. of an orange tree). Will try to clarify more in that paragraph!

File: clipseg-zero-shot.md [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+url = "https://unsplash.com/photos/8Nc_oQsc2qQ/download?ixid=MnwxMjA3fDB8MXxhbGx8fHx8fHx8fHwxNjcxMjAwNzI0&force=true&w=640"
+image = Image.open(requests.get(url, stream=True).raw)
+image
+```
+
+<figure class="image table text-center m-0 w-6/12">
+  <medium-zoom background="rgba(0,0,0,.7)" alt="Colab output" src="assets/123_clipseg-zero-shot/73d97c93dc0f5545378e433e956509b8acafb8d9.png"></medium-zoom>
+</figure>
+
+## Text prompting
```
[the remark this answers is not in the document]
Comment (segments-tobias) [reply clipseg-zero-shot.md:115]: Good catch

File: clipseg-zero-shot.md [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+
+_, ax = plt.subplots(1, len(prompts) + 1, figsize=(3*(len(prompts) + 1), 4))
+[a.axis('off') for a in ax.flatten()]
+ax[0].imshow(image)
+[ax[i+1].imshow(torch.sigmoid(preds[i][0])) for i in range(len(prompts))];
+[ax[i+1].text(0, -15, prompt) for i, prompt in enumerate(prompts)];
+```
+
+<figure class="image table text-center m-0 w-full">
+  <medium-zoom background="rgba(0,0,0,.7)" alt="Colab output" src="assets/123_clipseg-zero-shot/14c048ea92645544c1bbbc9e55f3c620eaab8886.png"></medium-zoom>
```
[the remark this answers is not in the document]
Comment (segments-tobias) [reply clipseg-zero-shot.md:149]: I'll properly add alt texts. I don't think the image names are very important, since they won't have to be touched anymore.

File: clipseg-zero-shot.md [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+</figure>
+
+The decoder is trained on the [PhraseCut dataset](https://arxiv.org/abs/2008.01187), which contains over 340,000 phrases with corresponding image segmentations. The authors also experimented with various augmentations to expand the size of the dataset. The goal here is not only to be able to segment the categories that are present in the dataset, but also to segment unseen categories. Experiments indeed show that the decoder can generalize to unseen categories.
+
+ 
+
+One interesting feature of CLIPSeg is that we input a CLIP embedding to specify which thing we want to segment. This CLIP embedding can come from a word or a piece of text, but also from another image. This means we can use an example image to segment objects that might be hard to describe (e.g. a specific logo). The authors call this “visual prompting”, and the paper contains some tips on improving the effectiveness of this technique. They find that cropping the query image so that it only contains the object you want to segment, helps a lot. Blurring and darkening the background of the query image also helps a little bit.
```
[the remark this answers is not in the document]
Comment (segments-tobias) [reply clipseg-zero-shot.md]: Let me know what you think of the new paragraphs I wrote.

**Comment by segments-tobias** [comment]
Thanks @sayakpaul for the review. I fixed your comments and created a PR for the assets.
Let me know if you have any other comments!

File: clipseg-zero-shot.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (sayakpaul) [reply clipseg-zero-shot.md:3]: Thank you so much!

You need to change the URLs accordingly.

[the remark this answers is not in the document]
Comment (segments-tobias) [reply clipseg-zero-shot.md:3]: Updated them now :)

**Review by sayakpaul — APPROVED** [review]
Awesome! LGTM from my end.

File: clipseg-zero-shot.md [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+    </a>
+</div>
+
+<script async defer src="https://unpkg.com/medium-zoom-element@0/dist/medium-zoom-element.min.js"></script>
+
+<a target="_blank" href="https://colab.research.google.com/github/huggingface/blog/blob/main/notebooks/123_clipseg-zero-shot.ipynb">
+    <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
+</a>
+
+**This guide shows how you can use [CLIPSeg](https://arxiv.org/abs/2112.10003), a zero-shot image segmentation model, using [`🤗 transformers`](https://huggingface.co/transformers). CLIPSeg creates rough segmentation masks that can be used for robot perception, image inpainting, and many other tasks. If you need more precise segmentation masks, we’ll show how you can refine the results of CLIPSeg on [Segments.ai](https://segments.ai/?utm_source=hf&utm_medium=blog&utm_campaign=clipseg).**
```
Comment (NielsRogge) [inline clipseg-zero-shot.md:41]:
```suggestion
**This guide shows how you can use [CLIPSeg](https://huggingface.co/docs/transformers/main/en/model_doc/clipseg), a zero-shot image segmentation model, using [`🤗 transformers`](https://huggingface.co/transformers). CLIPSeg creates rough segmentation masks that can be used for robot perception, image inpainting, and many other tasks. If you need more precise segmentation masks, we’ll show how you can refine the results of CLIPSeg on [Segments.ai](https://segments.ai/?utm_source=hf&utm_medium=blog&utm_campaign=clipseg).**
```

File: clipseg-zero-shot.md [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+
+## CLIP: the magic model behind CLIPSeg
+
+[CLIP](https://openai.com/blog/clip/), which stands for **C**ontrastive **L**anguage–**I**mage **P**re-training, is a model developed by OpenAI in 2021. You can give CLIP an image or a piece of text, and CLIP will output an abstract *representation* of your input. This abstract representation, also called an *embedding*, is really just a vector (a list of numbers). You can think of this vector as a point in high-dimensional space. CLIP is trained so that the representations of similar pictures and texts are similar as well. This means that if we input an image and a text description that fits that image, the representations of the image and the text will be similar (i.e., the high-dimensional points will be close together).
```
Comment (NielsRogge) [inline clipseg-zero-shot.md:51]:
```suggestion
[CLIP](https://huggingface.co/docs/transformers/main/en/model_doc/clip), which stands for **C**ontrastive **L**anguage–**I**mage **P**re-training, is a model developed by OpenAI in 2021. You can give CLIP an image or a piece of text, and CLIP will output an abstract *representation* of your input. This abstract representation, also called an *embedding*, is really just a vector (a list of numbers). You can think of this vector as a point in high-dimensional space. CLIP is trained so that the representations of similar pictures and texts are similar as well. This means that if we input an image and a text description that fits that image, the representations of the image and the text will be similar (i.e., the high-dimensional points will be close together).
```

File: clipseg-zero-shot.md [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+
+The reason why CLIP works so well is that the model was trained on a huge dataset of images with text captions. The dataset contained a whopping 400 million image-text pairs taken from the internet. These images contain a wide variety of objects and concepts, and CLIP is great at creating a representation for each of them.
+
+## CLIPSeg: image segmentation with CLIP
+
+[CLIPSeg](https://arxiv.org/abs/2112.10003) is a model that uses CLIP representations to create image segmentation masks. It was published by Timo Lüddecke and Alexander Ecker. They achieved zero-shot image segmentation by training a Transformer-based decoder on top of the CLIP model, which is kept frozen. The decoder takes in the CLIP representation of an image, and the CLIP representation of the the thing you want to segment. Using these two inputs, the CLIPSeg decoder creates a binary segmentation mask. To be more precise, the decoder doesn’t only use the final CLIP representation of the image we want to segment, but it also uses the outputs of some of the layers of CLIP.
```
Comment (NielsRogge) [inline clipseg-zero-shot.md:66]:
```suggestion
[CLIPSeg](https://arxiv.org/abs/2112.10003) is a model that uses CLIP representations to create image segmentation masks. It was published by Timo Lüddecke and Alexander Ecker. They achieved zero-shot image segmentation by training a Transformer-based decoder on top of the CLIP model, which is kept frozen. The decoder takes in the CLIP representation of an image, and the CLIP representation of the thing you want to segment. Using these two inputs, the CLIPSeg decoder creates a binary segmentation mask. To be more precise, the decoder doesn’t only use the final CLIP representation of the image we want to segment, but it also uses the outputs of some of the layers of CLIP. 
```

File: clipseg-zero-shot.md [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+  <medium-zoom background="rgba(0,0,0,.7)" alt="The mask of the coffee cup in the breakfast image." src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/123_clipseg-zero-shot/fbde45fc65907d17de38b0db3eb262bdec1f1784.png"></medium-zoom>
+</figure>
+
+Let's try one last time by using the visual prompting tips described in
+the paper, i.e. cropping the image and darkening the background.
+
+```python
+url = "https://i.imgur.com/mRSORqz.jpg"
+prompt_alt = Image.open(requests.get(url, stream=True).raw)
+prompt_alt
```
Comment (NielsRogge) [inline clipseg-zero-shot.md:204]:
```suggestion
alternative_prompt = Image.open(requests.get(url, stream=True).raw)
alternative_prompt
```

File: clipseg-zero-shot.md [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+  <medium-zoom background="rgba(0,0,0,.7)" alt="A cropped version of the image of the coffee cup with a darker background." src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/123_clipseg-zero-shot/915a97da22131e0ab6ff4daa78ffe3f1889e3386.png"></medium-zoom>
+</figure>
+
+```python
+encoded_prompt_alt = processor(images=[prompt_alt], return_tensors="pt")
+# predict
+with torch.no_grad():
+  outputs = model(**encoded_image, conditional_pixel_values=encoded_prompt_alt.pixel_values)
+preds = outputs.logits.unsqueeze(1)
+preds = torch.transpose(preds, 0, 1)
```
Comment (NielsRogge) [inline clipseg-zero-shot.md:217]:
```suggestion
encoded_alternative_prompt = processor(images=[alternative_prompt], return_tensors="pt")
# predict
with torch.no_grad():
  outputs = model(**encoded_image, conditional_pixel_values=encoded_alternative_prompt.pixel_values)
preds = outputs.logits.unsqueeze(1)
preds = torch.transpose(preds, 0, 1)
```

File: clipseg-zero-shot.md [hunk 11]
```diff
@@ -0,0 +1,391 @@
...
+
+As you can see, the results from CLIPSeg are a little fuzzy and very
+low-res. If we want to obtain better results, you can fine-tune a
+state-of-the-art segmentation model, as explained in [our previous
+blogpost](https://huggingface.co/blog/fine-tune-segformer). To finetune
+the model, we\'ll need labeled data. In this section, we\'ll show you
+how you can use CLIPSeg to create some rough segmentation masks and then
+refine them on
+[Segments.ai](https://segments.ai/?utm_source=hf&utm_medium=blog&utm_campaign=clipseg),
+the best labeling platform for image segmentation.
```
Comment (NielsRogge) [inline clipseg-zero-shot.md:245]: :D

File: clipseg-zero-shot.md [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+the biggest mistakes, and then you can use the corrected dataset to
+train a better model than CLIPSeg.
+
+<figure class="image table text-center m-0 w-9/12">
+  <medium-zoom background="rgba(0,0,0,.7)" alt="Thumbnails of the final segmentation labels on Segments.ai." src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/123_clipseg-zero-shot/segments-thumbs.png"></medium-zoom>
+</figure>
+
+## Conclusion
+
+CLIPSeg is a zero-shot segmentation model that works with both text and image prompts. The model adds a decoder to CLIP and can segment almost anything. However, the output segmentation masks are still very low-res for now, so you’ll probably still want to fine-tune a different segmentation model.
```
Comment (NielsRogge) [inline clipseg-zero-shot.md:389]:
```suggestion
CLIPSeg is a zero-shot segmentation model that works with both text and image prompts. The model adds a decoder to CLIP and can segment almost anything. However, the output segmentation masks are still very low-res for now, so you’ll probably still want to fine-tune a different segmentation model. Note that there's a lot of research going on to leverage CLIP for zero-shot segmentation, among which [GroupViT](https://huggingface.co/docs/transformers/model_doc/groupvit) which is also available in 🤗 Transformers. One can expect more models to be added in the near future.
```

**Review by NielsRogge — APPROVED** [review]
Awesome blog, thanks for all your work! 🙏

File: clipseg-zero-shot.md (hunk 11 above)
Comment (sayakpaul) [reply clipseg-zero-shot.md:245]: I forgot to come back to this one. 

@segments-tobias we usually try to avoid phrases like "best".

**Comment by sayakpaul** [comment]
@osanseviero 

@NielsRogge and I have reviewed and approved the blog post. Once the suggestions from Niels are reflected, I think the post can be safely merged and shipped.

File: clipseg-zero-shot.md (hunk 11 above)
Comment (NielsRogge) [reply clipseg-zero-shot.md:245]: No it's totally fine for me!

File: clipseg-zero-shot.md (hunk 12 above)
Comment (segments-tobias) [reply clipseg-zero-shot.md]: Rewrote this a little bit, can you check once more @NielsRogge ?

**Comment by segments-tobias** [comment]
I added the post to `_blog.yaml` and updated the publish date to today, so it should be good to go.

File: clipseg-zero-shot.md (hunk 11 above)
Comment (segments-tobias) [reply clipseg-zero-shot.md:245]: Lmk if you still want me to change it @sayakpaul

Comment (sayakpaul) [reply clipseg-zero-shot.md]: Appreciate your understanding. Thank you!

**Review by osanseviero — APPROVED** [review]
I checked that the blog format and directories are alright. As Niels and Sayak already reviewed, I'm approving!

File: clipseg-zero-shot.md (hunk 12 above)
Comment (NielsRogge) [reply clipseg-zero-shot.md]: LGTM 👍

**Comment by osanseviero** [comment]
https://huggingface.co/blog/clipseg-zero-shot cool blog! Congrats :fire:

**Comment by NielsRogge** [comment]
@segments-tobias could you open a PR to incorporate [this update](https://github.com/huggingface/transformers/pull/20816)?

**Comment by segments-tobias** [comment]
Done, see [https://github.com/huggingface/blog/pull/748](https://github.com/huggingface/blog/pull/748)
